yt_dlp/extractor/ard.py

```diff
@@ -577,3 +578,117 @@ def fetch_page(page_num):
+        },
+        'playlist_mincount': 22,
+    }, {
+        'url': 'https://www.ardaudiothek.de/episode/1live-caiman-club/caiman-club-s04e04-cash-out/1live/13556081/',
+        'info_dict': {
+            'id': '13556081',
+            'ext': 'mp3',
+            'upload_date': '20240717',
+            'duration': 3339,
+            'title': 'CAIMAN CLUB (S04E04): Cash Out',
+            'thumbnail': 'https://api.ardmediathek.de/image-service/images/urn:ard:image:d5014b612429c396',
+            'description': 'md5:8decf7974ed1cbf5a9d2c537940e1c4b',
+            'display_id': '1live-caiman-club/caiman-club-s04e04-cash-out/1live',
+            'timestamp': 1721181641,
+            'series': '1LIVE Caiman Club',
+
+        },
+    }]
+
+    _QUERY_PLAYLIST = '''show(id: "%s") {
+        title
+        description
+        items {
+            nodes {
+                url
+                episodeNumber
+                grouping
+                isPublished
+            }
+        }
+    }'''
+
+    _QUERY_ITEM = '''\
+    item(id: "%s") {
+        audioList {
+            href
+            distributionType
+        }
+        show {
+          title
+        }
+        image {
+          url
+        }
+        synopsis
+        title
+        duration
+        startDate
+    }'''
+
+    _GRAPHQL_ENDPOINT = 'https://api.ardaudiothek.de/graphql'
+
+    def _graphql_query(self, display_id, query):
+        return self._download_json(
+            self._GRAPHQL_ENDPOINT,
+            display_id,
+            data=json.dumps({'query': '{' + query + '}'}).encode(),
+            headers={
+                'Content-Type': 'application/json',
+            },
+        )['data']
+
+    def _real_extract(self, url):
+        video_id, display_id, playlist_type, season_number = self._match_valid_url(url).group(
+            'id', 'display_id', 'playlist', 'season')
+        if re.match('^[/-]*$', display_id):
+            display_id = video_id
+
+        if playlist_type:
+            playlist_info = self.graphql_query(display_id, self._QUERY_PLAYLIST % video_id)['show']
+            episodes = playlist_info['items']['nodes']
+            entries = []
+            for episode in episodes:
+                if episode['isPublished']:
+                    entries.append(self.url_result(
+                        episode['url'],
+                        ie=ARDAudiothekIE.ie_key()))
+            return self.playlist_result(entries, video_id, playlist_title=display_id, **traverse_obj(playlist_info, {
+                'title': ('title', {str}),
+                'description': ('description', {str}),
+            }))
+
+        item = self.graphql_query(display_id, self._QUERY_ITEM % video_id)['item']
+        audio_list = item.get('audioList', [])
+        return {
+            'display_id': display_id,
+            'formats': [{'url': x['href'], 'format_id': x['distributionType']} for x in audio_list],
```

Comment:
we can use traversal here as well:
```suggestion
        return {
            'display_id': display_id,
            'formats': traverse_obj(self.graphql_query(display_id, self._QUERY_ITEM % video_id), (
                'item', 'audioList', lambda _, v: url_or_none(v['href']), {
                    'url': 'href',
                    'format_id': ('distributionType', {str}),
            })),
```

Reply:
Thanks for all the feedback! In this case, I'm not sure I consider this to be more readable … but I can adapt :)

Reply:
Its not only about being more readable. While the convention is to try and avoid using variables only used once, this can be broken if readability suffer.

However, `traverse_obj` is defensive and will never cause a `KeyError`, `TypeError` or similar. While very strange for a graphql query to change shape, for various other apis this new format is preferred to allow a non fatal extraction. That way metadata can still be accessible eventhough formats are missing

Reply:
The variable `item` is not only used once here, the second occurrence is now causing a code check error:
> yt_dlp/extractor/ard.py:689:28: F821 Undefined name `item`

I had to reinstate the definition of this variable to make the extractor work again.